File: CortexSuite/vision/benchmarks/localization/src/c/readSensorData.c
/********************************
Author: Sravanthi Kota Venkata
********************************/

#include <stdio.h>
#include <stdlib.h>
#include "localization.h"

F2D* readSensorData(I2D* index, F2D* fid, I2D* type, I2D* eof)
{
    F2D *retData;
    int rows, i, j, k;
    int atype=-1, aindex;

    aindex = asubsref(index, 0);

    asubsref(index,0) = asubsref(index,0) + 1;
    rows = fid->height;
    asubsref(type,0) = 0;
    retData = fSetArray(1, 8, 0);
    
    if( asubsref(index,0) > (rows-1) )
        asubsref(eof,0) = 1;
    else
    {
        if( asubsref(index,0) == rows)
            asubsref(eof,0) = 1;
        else
            asubsref(eof,0) = 0;

        k = asubsref(index,0);
        atype = subsref(fid, k, 1);
        if( (atype == 1) || (atype == 2) || (atype == 3) )
        {
            for(i=0; i<3; i++)
            {
                asubsref(retData,i) = subsref(fid,k,(i+2));
            }
        }
        if( atype == 4 )
        {
            for(i=0; i<3; i++)
            {
                asubsref(retData,i) = subsref(fid,k,(i+2));
            }
            for(i=3; i<8; i++)
            {
                asubsref(retData,i) = subsref(fid,k+1,(i-3));
            }
            aindex = aindex + 1;
        }
        aindex = aindex + 1;
    }

    asubsref(index,0) = aindex;
    asubsref(type, 0) = atype;

    return retData;
}

File: CortexSuite/vision/benchmarks/localization/src/matlab/readSensorData.m
function [retTStamp, retType, retData, retEOF, index]=readSensorData(index1, fid)

index = index1+1;
rows = size(fid, 1);
retTStamp = 0;
retType = 0;
retData = zeros(1,8);

if(index > rows)
    retEOF = 1;
else
%     for i=index:rows
%         index = i;
%         if(fid(i,2) == 4)
%             break;
%         end
%     end
    if(index == rows)
        retEOF = 1;
    else
        retEOF = 0;
    end

    k = index;
    retTStamp=fid(k,1);
    retType=fid(k,2);
    if(fid(k, 2) == 1 || fid(k, 2) == 2 || fid(k, 2) == 3)
        index = k;
        for i=1:3
            retData(1,i)=fid(k,i+2);
%            fprintf(1,'retData,i -> %f\t%d\n', retData(1,i), i);
        end
    end
    if(fid(k, 2) == 4)
        index = k;
        for i=1:3
            retData(1,i)=fid(k,i+2);
%            fprintf(1,'retData,i -> %f\t%d\n', retData(1,i), i);
        end
        for i=4:8
            retData(1,i) = fid(k+1,i-3);
%            fprintf(1,'retData,i -> %f\t%d\n', retData(1,i), i);
        end
        index = index + 1;        
    end    
end
